api/include/opentelemetry/trace/trace_state.h

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <cstring>
+
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/nostd/unique_ptr.h"
+
+namespace opentelemetry
+{
+namespace trace
+{
+
+/**
+ * TraceState carries tracing-system specific context in a list of key-value pairs. TraceState
+ * allows different vendors to propagate additional information and inter-operate with their legacy
+ * Id formats.
+ *
+ * Implementation is optimized for a small list of key-value pairs.
+ *
+ * Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+ * and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+ * forward slashes /.
+ *
+ * Value is opaque string up to 256 characters, of printable ASCII RFC0020 characters (i.e. the
+ * range 0x20 to 0x7E) except comma , and equals =.
+ */
+class TraceState
+{
+public:
+  static constexpr int kKeyMaxSize       = 256;
+  static constexpr int kValueMaxSize     = 256;
+  static constexpr int kMaxKeyValuePairs = 32;
+
+  // Class to store key-value pairs.
+  class Entry
+  {
+  public:
+    Entry() noexcept = default;
+
+    // Creates an Entry for a given key-value pair.
+    Entry(nostd::string_view key, nostd::string_view value) noexcept;
+
+    const char *GetKey();
+    const char *GetValue();
+
+  private:
+    // Store key and value as raw char pointers to avoid using std::string.
+    const char *key_;
+    const char *value_;
+  };
+
+  // An empty TraceState.
+  TraceState() noexcept : entries_(new Entry[kMaxKeyValuePairs]), num_entries_(0) {}
+
+  // Returns false if no such key, otherwise returns true and populates value.
+  bool Get(nostd::string_view key, nostd::string_view value) const noexcept { return false; }
+
+  // Creates an Entry for the key-value pair and adds it to entries.
```

Comment:
Need to clarify, if value is null, would this remove the entry or not, and do we need a separate function to delete a key-value pair.

Reply:
Take a look at this https://www.w3.org/TR/trace-context/#mutating-the-tracestate-field.

Reply:
For deletion, it says `Vendors SHOULD NOT delete keys that were not generated by them. The deletion of an unknown key/value pair will break correlation in other systems.` If we have a deletion function, how will we ensure that we're only deleting entries generated by the current vendor?

Reply:
I guess it is the vendors' problem not ours?

Reply:
Would it be possible to limit the scope and go without the ability to delete entries? We could always add it later on.

Reply:
Yeah, not blocking the PR.

Reply:
If `value` is null, it would make sense to me for this to be a no-op, rather than deleting the entry.

Reply:
@reyang @pyohannes I added a comment for this, please let me know your thoughts!